app/gui2/src/components/visualizations/TableVisualization.vue

```diff
@@ -135,11 +136,25 @@ const selectableRowLimits = computed(() => {
 })
 const wasAutomaticallyAutosized = ref(false)
 
-const numberFormat = new Intl.NumberFormat(undefined, {
-  style: 'decimal',
-  maximumFractionDigits: 12,
-  useGrouping: 'min2' as any,
-})
+function isGrouingRequired(params: ICellRendererParams) {
+  const field = params.colDef?.field
+  const allDataForCol = params.node.parent?.allLeafChildren.map((row) => {
+    const data = row.data[field || '']
+    return data?.type === 'BigInt' ? data.value : data
+  })
+  return allDataForCol?.find((num) => num > 9999) != null
```

Comment:
This is a sequential scan across all table rows every time we are formatting any of those column's numbers. It has complexity `O(n^2)` for a very frequent operation on potentially very large datasets. This needs to be done differently, or we risk very big stutters whenever the visualization is visible and interacted with.

I'd be fine with performing a single pass scan over whole dataset whenever it changes, and save the computed formatting info per column. You can use vue's `computed` to do that computation once for the whole dataset, then reuse it in `formatNumber`.

Also, the `array.find(predicate) != null` can be simplified to `array.some(predicate)`.

Reply:
I have Changed this so it only scans the data once, let me know what you think!